### DISCUSSION TITLE
Watch mode by default

### DISCUSSION THREAD

**Discussion by EvHaus** [vitest-dev/vitest#540] [Q&A] [answered]
At the moment, watch mode is enabled by default for Vitest. What does everyone think about that?

Personally, I found it more troublesome than helpful. I like creating npm scripts for my test runners, so I have `yarn test` configured to run `vitest`. I would prefer watch mode to be off by default, but if I change the vitest configuration to `{watch: false}` now I can't turn it back on with `yarn test --watch=true` anymore.

Anyone else have a similar challenge?

**Comment by charles-allen** [comment]
Have you seen the `--run` option?

I was also initially confused, partly because glancing at the docs gives the wrong impression it's a separate command `vitest watch` (you need to read the text to realise the watch part does nothing), but mostly because every other similar CLI that I've used is single-run by default with watch as an option.

I would keep watch by default & improve the docs

**Reply by EvHaus** [reply]
Ah, good idea. `yarn vitest --run` is indeed a good alternative here.

I would still advocate for making `run` the default (as opposed to `watch`), which would make it more consistent with Jest's behaviour as well. But now that I have a solution, it's just a subjective opinion.

**Comment by antfu** [comment, marked as answer]
Yes, that's an intentional decision as we would like to take the chance to challenge the default we all used to. It might feel a bit confusing at the beginning if you are switching from Jest, but if you think that Vite is more like a dev server, and we provide smart test running based on module graph, having watch by default actually make more sense for development (in CI it will be run mode). We are trying to make the migration easier from Jest, but that doesn't necessarily mean we need to follow every decision that Jest did.

**Reply by ivancuric** [reply]
How does it know when it runs in CI mode?

**Reply by cyyynthia** [reply]
> How does it know when it runs in CI mode?

It checks if the `CI` env variable is present, or if it can detect a known environment variable from a list of providers. The former is enough for the majority of providers, with AWS Amplify, Cloudflare Pages, Cloudflare Workers CI/CD, Codesphere, and Firebase App Hosting being explicitly checked.

- https://github.com/unjs/std-env/blob/7e8cb7b1000f313a03d7645c82e2d8608c250dd6/src/flags.ts#L9
- https://github.com/unjs/std-env/blob/7e8cb7b1000f313a03d7645c82e2d8608c250dd6/src/providers.ts#L60-L115

**Comment by will-stone** [comment]
A use case for _run by default_ (or at least a way to switch it via config) is in a monorepo of other tools that default to _run_. For example, if you run `npm test -ws --if-present`, all but Vite projects will run the `test` script which is _run once only_. Then if you wanted to run just one in watch mode you can (without checking for other script names) run `npm test -w WORKSPACE_NAME -- --watch`, or just one in _run once mode_ by running `npm test -w WORKSPACE_NAME`. By this being the opposite in Vite, this consistency is lost.
Great project though, thank you for all the hard work. I'm loving the in-source tests for a util lib I'm building 😄

**Reply by will-stone** [reply]
Oh, my bad, this _is_ possible 😖 

```js
export default defineConfig({
  test: {
    watch: false,
  },
})
```

**Reply by tokland** [reply]
Cool! For completeness, the header is: 

```typescript
/// <reference types="vitest" />
import { defineConfig } from "vite";
```

**Reply by connorjs** [reply]
Using `watch: false` in the Vite(st) config is my current solution for Nx monorepo (@nx/vite changed to watch by default in local in 19.1.0).

Traditionally, I have had `test` command that defaults to watch (nice to use `npm t` on CLI) and skips coverage. I then have a `coverage` command that runs and collects coverage. CI builds use `coverage`, not `test`.
